### DISCUSSION TITLE
Software Serial Bridge

### DISCUSSION THREAD

**Discussion by hamster65** [arendst/Tasmota#10203] [Q&A] [answered]
Hi,
I have got a wemos D1 module with firmware binary tasmota-sensors (needed for SenseAir S8 sensor). I want to receive serial data with software serial bridge on Pin D0. I configured D0 to SerBr RX. I defined no TX pin because I don't need to send data.

But all software serial commands like SBaudrate or SSerialSend are not available (unknown command).
Is software serial bridge not available with the tasmota-sensors binary? It is available in the drop down menu for the GPIO assignments....

Thanks!

**Comment by sfromis** [comment]
You very much need the TX pin for Senseair S8, the protocol includes sending commands to the sensor. Additionally, the driver will simply not be active unless both needed pins are configured.

**Reply by hamster65** [reply]
Hm. I configured RX and TX pins for software serial in the meantime, still does not work. I use different pins for the co2 sensor (which works fine). I just cannot issue any software serial commands.

**Reply by sfromis** [reply]
I've only used Senseair S8 with the regular hardware serial, but I do not see how it could make sense for you to try to use the software serial bridge commands when the serial pins have to be configured specifically for the Senseair S8 sensor, not generic software serial. The driver for Senseair S8 will take care of whatever config is needed. It can only be consistent with software serial bridge if you do not use the Tasmota support for the sensor, but try to do it with your own serial communication. To me it seems like Tasmota will use software serial to talk to the sensor if you do not run it on gpio 1/3 or 13/15, but still config the appropriate gpio pins for the sensor.

**Comment by hamster65** [comment, marked as answer]
I solved it: I use the hardware rx pin to receive communication from my device (I had to desolder the usb uart chip for this to work because it pulled up my rather weak serial line). the S8 sensor is connected to different pins.

But even if I undefine the sensor pins and add only serial bridge rx and tx, it always sends and receives using the hardware rx and tx pins and does not recognize the SSerial-Commands.

**Reply by Frogmore42** [reply]
If I understand correctly, you actually have two serial devices, the S8 and something else.  The S8 will use software serial when it is not connected to the HW serial pins.  I don't know if software serial can handle multiple copies, ie two software serial channels.  It sounds like that is what you needed and wanted, but it didn't work.  You solved it by using HW serial for your other device.

**Reply by hamster65** [reply]
Yes. But even when I did not set up pins for the CO2 sensor but only for software serial bridge, I would not let me use software serial commands.

**Reply by walkingsnail** [reply]
I met a quite similar problem, and would like to see if can have any suggestion here.

I am using a board with ESP32s2, and flashed a Tasmota 10.1.0 firmware.

I connected two sensors to ESP, one MH-Z19B CO2 sensor connected to GPIO 4/3 and another PMS7003 PM2.5 sensor connected to pin GPIO 7/8. I also tested other GPIO like 17/18/40/41, the same result)

When I configured either of them (TX/RX) in Tasmota, it works perfectly. However, if I configure both of them, only CO2 sensor works, and the command "sensor18" results error.

Appreciate any feedback

**Reply by barbudor** [reply]
@walkingsnail different ESP, different problem
See https://github.com/arendst/Tasmota/issues/14487#event-5907361770
As of today Tasmota doesn't have software Serial on any ESP32. Only 1 serial available for sensors on C3 and S2 but 2 on ESP32.
If you need 2 serial sensors, move to standard ESP32
No solutions in short term but we are thinking about it.

**Reply by walkingsnail** [reply]
Thank you, Barbudor. But why S2 can't support 2 serial for sensors if we disable serial logging? There are 2 serials in S2.

I checked "TasmotaSerial.cpp", change "if (tasmota_serial_index > 0)" to ">= 0", and set serial logging to 0, then find sometimes both sensors work! Each time after reset, both sensors' information will be displayed in web UI, after ~30s, only MHZ-19 appears. Later, both sensors' info are available from time to time, very rare, I still didn't find the law when we have both sensors' info.

I am not a programmer, just understand some basic source code logic. It seems above change enabled UART0 to be used for sensor, but other codes disabled it later, or have conflict.

Doese it make sense that it is possible S2 can support two sensors if we doesn't need serial logging at all?

--------------------------------------------------------------
Below is what I see in web logging, you will see both sensors' result at 06:27:43:199, and later there are some strange commands, then at 06:32:43.072 only MHZ-19 data is reported:

6:27:42.697 MQT: tele/air_station/STATE = {"Time":"2022-01-19T06:27:42","Uptime":"0T00:00:27","UptimeSec":27,"Heap":72,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"EF","BSSId":"88:C3:97:E6:F4:36","Channel":2,"Mode":"11n","RSSI":80,"Signal":-60,"LinkCount":1,"Downtime":"0T00:00:19"}}
06:27:42.946 MQT: domoticz/in = {"idx":12,"nvalue":529,"Battery":100,"RSSI":8}
06:27:43.072 MQT: domoticz/in = {"idx":13,"nvalue":0,"svalue":"17","Battery":100,"RSSI":8}
06:27:43.199 MQT: tele/air_station/SENSOR = {"Time":"2022-01-19T06:27:42","MHZ19B":{"Model":"B","CarbonDioxide":529,"Temperature":22.0},"PMS5003":{"CF1":15,"CF2.5":17,"CF10":18,"PM1":15,"PM2.5":17,"PM10":18,"PB0.3":2862,"PB0.5":850,"PB1":41,"PB2.5":2,"PB5":2,"PB10":2},"ESP32":{"Temperature":36.0},"TempUnit":"C"}
06:27:43.502 CMD: BMpttJMNS104
06:27:43.511 MQT: stat/air_station/RESULT = {"Command":"Unknown"}
06:27:44.477 CMD: #2
06:27:57.894 CMD: LBM
06:27:57.901 MQT: stat/air_station/RESULT = {"Command":"Unknown"}

......

06:32:42.687 MQT: tele/air_station/STATE = {"Time":"2022-01-19T06:32:42","Uptime":"0T00:05:27","UptimeSec":327,"Heap":71,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"EF","BSSId":"88:C3:97:E6:F4:36","Channel":2,"Mode":"11n","RSSI":90,"Signal":-55,"LinkCount":1,"Downtime":"0T00:00:19"}}
06:32:42.938 MQT: domoticz/in = {"idx":12,"nvalue":568,"Battery":100,"RSSI":9}
06:32:43.072 MQT: tele/air_station/SENSOR = {"Time":"2022-01-19T06:32:42","MHZ19B":{"Model":"B","CarbonDioxide":568,"Temperature":21.0},"ESP32":{"Temperature":36.0},"TempUnit":"C"}

**Reply by walkingsnail** [reply]
one more info: now 'sensor18' command have normal result:

06:49:34.418 MQT: stat/air_station/RESULT = {"Sensor18":0}

**Reply by barbudor** [reply]
The basic logic is there
When ESP32 started, with it's 3 serial it looked reasonnable to lock 1 serial for the console
Now that C3/S2 are in between the ESP8266 and ESP32, we need to think
Making UART0 available for sensors is a solution
Porting SoftwareSerial to ESP32 is another
This has to be evaluated

**Reply by arendst** [reply]
Pls try latest development version where I enable uart0 for other usage disabling serial logging.

It's experimental so pls report back your findings.

**Reply by ortegafernando** [reply]
Hi, is there any option to enable two software serial bridges in esp8266 ? thanks

**Reply by sfromis** [reply]
No. But with ESP32 and Berry, it is possible to have more than one serial; still not the serial bridge.

**Reply by ortegafernando** [reply]
> No. But with ESP32 and Berry, it is possible to have more than one serial; still not the serial bridge.

Thanks, so there is no "software" solution (or easy solution) to have 2 software serial ports in esp8266, isn't it? I can use hardware serial brdige + software serial bridge in esp8266, can't I?
Thanks

**Reply by sfromis** [reply]
I'm not 100% sure, but my expectation would be no. The thing is that the places in Tasmota where you use serial does not have a way of selecting which. Maybe scripting can allow pin selection, but that's an area I know very little about.

**Reply by barbudor** [reply]
You should be able to use serial bridge and software Serial bridge separately. They have different commands set. But also different featurs/capabilities
